### DISCUSSION TITLE
Harmonising targets and workflow R packages

### DISCUSSION THREAD

**Discussion by jameslairdsmith** [ropensci/targets#463] [Ideas]
In my mind, targets and R packages are currently very different ways to go about creating a workflow projects. This is a shame, because in my work I find myself wanting to use aspects of both. I want to the visualisation, caching, dependency tracking etc. of targets, but also the well-established conventions, tooling and distribution mechanisms of R packages. I know there were attempts, like with [drakepkg](https://github.com/tiernanmartin/drakepkg), to bring them together, but those seem to have stopped? And there are (understandable) warnings in place when using things like `devtools::load_all()`.

If some kind of integration could be created, it would be amazing. But to be successful my guess is it would also need to be easy and seamless enough that users didn't have to invest too much effort into making it work. I've been thinking a bit about how this could be done, but I'm afraid I don't have many good answers.

One thing that did come to mind was perhaps using the roxygen2 infrastructure as a spring board for defining targets. There could perhaps be one or more custom tags; something like `#' @target`, which would designate certain functions as being part of a targets pipeline. These could then programatically generate a targets pipeline much in the same way `devtools::document()` generates documentation.

**Comment by wlandau** [comment]
> In my mind, targets and R packages are currently very different ways to go about creating a workflow projects. This is a shame, because in my work I find myself wanting to use aspects of both. I want to the visualisation, caching, dependency tracking etc. of targets, but also the well-established conventions, tooling and distribution mechanisms of R packages. I know there were attempts, like with drakepkg, to bring them together, but those seem to have stopped? And there are (understandable) warnings in place when using things like devtools::load_all().

The guidance at https://books.ropensci.org/targets/practices.html#packages-based-invalidation is relevant here. A common pattern is to implement most of the workflow's custom functions in a separate package, install that package locally, create a `targets` pipeline in a different directory, and then track the package functions using `tar_option_set(packages = "yourpkg", imports = "yourpkg")`. I often do this at work when I write an R package interface around a Bayesian model and then create a separate `targets` pipeline to [validate](https://docs.ropensci.org/stantargets/articles/mcmc_rep.html) it. If you want, you can go a step further and write a function to generate a `_targets.R` file for your pipeline (possibly using `tar_script()` or `tar_helper()`). Then, you have a portable mechanism for instantiating your pipeline in multiple places (hopefully avoiding writing the `_targets/` data store to the package source files).

R packages are a [debatable](https://www.milesmcbain.com/posts/an-okay-idea/) mechanism for managing one-off instances of project-oriented workflows. They are reliable, structured, and standardized, but they come with requirements and behaviors that do not always suit [research compendia](https://github.com/benmarwick/rrtools). But if you plan to instantiate multiple independent instances of the pipeline in production, the package approach sounds excellent.

> One thing that did come to mind was perhaps using the roxygen2 infrastructure as a spring board for defining targets. There could perhaps be one or more custom tags; something like #' @target, which would designate certain functions as being part of a targets pipeline. These could then programatically generate a targets pipeline much in the same way devtools::document() generates documentation.

Yeah, this feels similar to https://github.com/ropensci/tarchetypes/discussions/48. Seems like there would be significant gains in modularity and documentation. Right now, I recommend implementing this through [target factories](https://wlandau.github.io/targetopia/contributing.html#target-factories) and ordinary `roxygen2` docstrings, an approach that [other packages](https://wlandau.github.io/targetopia/packages.html) have successfully used to extend `targets`. Please let me know if there is something specific a custom `@target` tag could do over and above the usual R Targetopia approach.
